test/unit/bitmap_test.c

```diff
@@ -0,0 +1,162 @@
+ */
+bitmap test_alloc(heap h) {
+    bitmap b = allocate_bitmap(h, h, infinity);
+    bitmap_foreach_set(b, i) {
+        if (i) {
+            msg_err("!!! allocation failed for bitmap\n");
+            return NULL; 
+        }
+    }
+    return b;
+}
+
+/**
+ *  Tests cloning of bitmap using bitmap_clone 
+ *  function.
+ *  Also implicity tests bitmap_foreach_set
+ */  
+boolean test_clone(bitmap b) { 
+    for (int i = 0; i < 20; i++) 
+        bitmap_set(b, rand(), 1);
+    bitmap b_cpy = bitmap_clone(b);
+    bitmap_foreach_set(b, j) {
+        // implicit test for bitmap_foreach_set
+        if (!bitmap_get(b, j)) {
+            msg_err("!!! foreach_set failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+        if ((bitmap_base(b)[j >> 6] & (1ull << (j & 63))) != 
+            (bitmap_base(b_cpy)[j >> 6] & (1ull << (j & 63)))) {
+            msg_err("!!! cloning failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+    }
+    deallocate_bitmap(b_cpy);
+    return true;
+}
+
+/**
+ *  Tests copying of bitmap using bitmap_copy
+ *  function.
+ *  Also implicity tests bitmap_foreach_set
+ */ 
+boolean test_copy(heap h, bitmap b) { 
+    bitmap b_cpy = allocate_bitmap(h, h, infinity);
+    for (int i = 0; i < 20; i++) {
+        bitmap_set(b_cpy, rand(), 1);
+        bitmap_set(b, rand(), 1);
+    }
+    bitmap_copy(b, b_cpy);
+    bitmap_foreach_set(b, j) {
+        // implicit test for bitmap_foreach_set
+        if (!bitmap_get(b, j)) {
+            msg_err("!!! foreach_set failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+        if ((bitmap_base(b)[j >> 6] & (1ull << (j & 63))) != 
+            (bitmap_base(b_cpy)[j >> 6] & (1ull << (j & 63)))) {
+            msg_err("!!! copying failed for bitmap\n");
+            deallocate_bitmap(b_cpy);
+            return false;
+        }
+    }
+    deallocate_bitmap(b_cpy);
+    return true;
+}
+
+/**
+ *  Tests getting and setting bits of a bitmap using 
+ *  bitmap_get and bitmap_set functions.
+ */ 
+boolean test_set_and_get(bitmap b) {
+    u64 i = rand();
+    bitmap_set(b, i, 1);
+    if (!bitmap_get(b, i)) {
+        msg_err("!!! set and get failed for bitmap\n");
+        return false;
+    }
+    return true;
+}
+
+/**
+ *  Tests wrapping and unwrapping of bitmap using 
+ *  bitmap_wrap and bitmap_unwrap functions.
+ */ 
+boolean test_wrap(heap h) {
+    u64 map = rand();
+    bitmap b = bitmap_wrap(h, &map, infinity);
+    if (b->alloc_map->contents != &map) {
+        msg_err("!!! wrap failed for bitmap\n");
+        bitmap_unwrap(b);
+        return false;
+    }
+    bitmap_unwrap(b);
+    return true;
+}
+
+boolean test_bitmap_alloc(bitmap b, u64 start, u64 end) {
+    u64 nbits = rand();
+    u64 first = bitmap_alloc_within_range(b, nbits, start, end);
+    if (first != INVALID_PHYSICAL && !bitmap_dealloc(b, first, nbits)) {
```

Comment:
`nbits` should not be larger than (end - start), otherwise bitmap_alloc_within_range() fails and then bitmap_dealloc() is not tested.
So if `nbits` has an appropriate value, the above `if` can be replaced by `if ((first == INVALID_PHYSICAL) || !bitmap_dealloc(b, first, nbits))`

Reply:
My test for bitmap_alloc within a specified range (line 187) is failing because bitmap_alloc_internal fails the first condition check on line 102 (bitmap.c). 

I attempted to fix this issue by recalculating nbits until it satisfies a condition that wouldn't fail bitmap_alloc (lines 145-153 in bitmap_test.c), but this isn't an efficient solution.

Do you have any suggestions of how I can ensure a combination of _start, end,_ and _nbits_ will lead to a successful bitmap_alloc?

Reply:
OK, I see now that a bit range is always allocated starting from a bit that is multiple of the range length, so as you noted having `nbits <= (end - start)` doesn't guarantee a successful allocation.
I wouldn't make the test code dependent on this internal implementation detail of the bitmap code (which could potentially change in the future), so I would just start with `nbits = rand()` and then call bitmap_alloc_within_range() until it succeeds, halving the `nbits` value after each failed attempt, as in:
```
while (true) {
    first = bitmap_alloc_within_range(b, nbits, start, end);
    if (first != INVALID_PHYSICAL)
        break;
    else
        nbits >>= 1;
}
```

Reply:
Actually, I realized that I was incorrectly calculating `nbits` in the first place. I was generating a random number between `start `and `end` for `nbits` instead of a random number between 0 and (end-start).

Reply:
> Actually, I realized that I was incorrectly calculating `nbits` in the first place. I was generating a random number between `start `and `end` for `nbits` instead of a random number between 0 and (end-start).

Well, the fact that this test is passing now doesn't guarantee that it will pass with other random values. And if you try calling test_bitmap_alloc() many times with different random values, one of these calls will end up failing, because of the fact that a bit range is always allocated starting from a bit that is multiple of the range length. The test should be robust and not fail, regardless of how many times it's executed.

Reply:
I see, thanks for clarifying that. 
In that case, if we loop until `nbits` satisfies `bitmap_alloc_within_range`, there wouldn't be an opportunity for this test to fail, right? So there isn't any error checking necessary when testing this function?

Reply:
There is some more error checking that can be added, such as:
- if (first != INVALID_PHYSICAL), checks that nbits is <= (end - start)
- check that all bits in the allocated range are actually set (similarly to what is done in `test_range_check_set()`)
- at the end, call bitmap_dealloc() and check that it returns true